XenModel/Actions/Updates/DownloadAndUpdateClientAction.cs

```diff
@@ -0,0 +1,286 @@
+            }
+
+            //if this is still the case after having retried MAX_NUMBER_OF_TRIES number of times.
+            if (updateDownloadState == DownloadState.Error)
+            {
+                log.ErrorFormat("Giving up - Maximum number of retries ({0}) has been reached.", MAX_NUMBER_OF_TRIES);
+                throw updateDownloadError ?? new Exception(Messages.ERROR_UNKNOWN);
+            }
+        }
+
+        private void NetworkAvailabilityChanged(object sender, NetworkAvailabilityEventArgs e)
+        {
+            if (!e.IsAvailable && client != null && updateDownloadState == DownloadState.InProgress)
+            {
+                updateDownloadError = new WebException(Messages.NETWORK_CONNECTIVITY_ERROR);
+                updateDownloadState = DownloadState.Error;
+                client.CancelAsync();
+            }
+        }
+
+        protected override void Run()
+        {
+            if (downloadUpdate)
+            {
+                log.InfoFormat("Downloading '{0}' installer (from '{1}') to '{2}'", updateName, address, outputPathAndFileName);
+                Description = string.Format(Messages.DOWNLOAD_CLIENT_INSTALLER_ACTION_DESCRIPTION, updateName);
+                LogDescriptionChanges = false;
+                DownloadFile();
+                LogDescriptionChanges = true;
+
+                if (IsCompleted || Cancelled)
+                    return;
+
+                if (Cancelling)
+                    throw new CancelledException();
+            }
+            
+            ValidateMsi();
+
+            if (!File.Exists(outputPathAndFileName))
+                throw new Exception(Messages.DOWNLOAD_CLIENT_INSTALLER_MSI_NOT_FOUND);
+
+            // Install the msi            
+            try
+            {
+                // Start the install process, it will handle closing of application.
+                Process.Start(outputPathAndFileName);
+                log.DebugFormat("Update {0} found and install started", updateName);
+            }
+            catch (Exception e)
+            {
+                if (File.Exists(outputPathAndFileName))
+                    File.Delete(outputPathAndFileName);
+
+                log.Error("Exception occurred when starting the installation process.", e);
+                throw;
+            }
+
+            Description = Messages.COMPLETED;
+        }
+
+        private void ValidateMsi()
+        {
+            using (FileStream stream = new FileStream(outputPathAndFileName, FileMode.Open, FileAccess.Read))
+            {
+                var calculatedChecksum = string.Empty; 
+
+                var hash = StreamUtilities.ComputeHash(stream, out _);
+                if (hash != null)
+                    calculatedChecksum = string.Join("", hash.Select(b => $"{b:x2}"));
+
+                // Check if calculatedChecksum matches what is in chcupdates.xml
+                if (checksum.ToLower() != calculatedChecksum.ToLower())

```

Comment:
The culture specific casing rules can be unexpected sometimes, although hex digits are probably pretty safe.
If you want to be really paranoid about avoiding cultural differences in string comparison I'd recommend going for either:
```csharp
if (checksum.ToLowerInvariant() != calculatedChecksum.ToLowerInvariant())
```
Or
```csharp
if (!checksum.Equals(calculatedChecksum, StringComparison.InvariantCultureIgnoreCase))
```
But I wouldn't block the PR on it.

Reply:
See also: http://www.moserware.com/2008/02/does-your-code-pass-turkey-test.html